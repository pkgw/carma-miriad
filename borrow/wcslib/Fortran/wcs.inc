*=======================================================================
*
* WCSLIB 4.18 - an implementation of the FITS WCS standard.
* Copyright (C) 1995-2013, Mark Calabretta
*
* This file is part of WCSLIB.
*
* WCSLIB is free software: you can redistribute it and/or modify it
* under the terms of the GNU Lesser General Public License as published
* by the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* WCSLIB is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
* License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with WCSLIB.  If not, see http://www.gnu.org/licenses.
*
* Direct correspondence concerning WCSLIB to mark@calabretta.id.au
*
* Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
* http://www.atnf.csiro.au/people/Mark.Calabretta
* $Id$
*=======================================================================

*     Functions.
      EXTERNAL  WCSCOPY, WCSFREE, WCSGET, WCSGTC, WCSGTD, WCSGTI,
     :          WCSINI, WCSMIX, WCSNPS, WCSNPV, WCSP2S, WCSPERR, WCSPRT,
     :          WCSSPTR, WCSPTC, WCSPTD, WCSPTI, WCSPUT, WCSS2P, WCSSET,
     :          WCSSUB
      INTEGER   WCSCOPY, WCSFREE, WCSGET, WCSGTC, WCSGTD, WCSGTI,
     :          WCSINI, WCSMIX, WCSNPS, WCSNPV, WCSP2S, WCSPERR, WCSPRT,
     :          WCSSPTR, WCSPTC, WCSPTD, WCSPTI, WCSPUT, WCSS2P, WCSSET,
     :          WCSSUB

*     Length of the WCSPRM data structure (INTEGER array) on 64-bit
*     machines.  Only needs to be 416 on 32-bit machines.
      INTEGER   WCSLEN
      PARAMETER (WCSLEN = 474)

*     Codes for WCS data structure elements used by WCSPUT and WCSGET.
      INTEGER   WCS_ALT, WCS_ALTLIN, WCS_CD, WCS_CDELT, WCS_CNAME,
     :          WCS_COLAX, WCS_COLNUM, WCS_CRDER, WCS_CROTA, WCS_CRPIX,
     :          WCS_CRVAL, WCS_CSYER, WCS_CTYPE, WCS_CUNIT, WCS_DATEAVG,
     :          WCS_DATEOBS, WCS_EQUINOX, WCS_FLAG, WCS_LATPOLE,
     :          WCS_LONPOLE, WCS_MJDAVG, WCS_MJDOBS, WCS_NAXIS, WCS_NPS,
     :          WCS_NPSMAX, WCS_NPV, WCS_NPVMAX, WCS_OBSGEO, WCS_PC,
     :          WCS_PS, WCS_PV, WCS_RADESYS, WCS_RESTFRQ, WCS_RESTWAV,
     :          WCS_SPECSYS, WCS_SSYSOBS, WCS_SSYSSRC, WCS_VELANGL,
     :          WCS_VELOSYS, WCS_VELREF, WCS_WCSNAME, WCS_ZSOURCE

      PARAMETER (WCS_FLAG     = 100)
      PARAMETER (WCS_NAXIS    = 101)
      PARAMETER (WCS_CRPIX    = 102)
      PARAMETER (WCS_PC       = 103)
      PARAMETER (WCS_CDELT    = 104)
      PARAMETER (WCS_CRVAL    = 105)
      PARAMETER (WCS_CUNIT    = 106)
      PARAMETER (WCS_CTYPE    = 107)
      PARAMETER (WCS_LONPOLE  = 108)
      PARAMETER (WCS_LATPOLE  = 109)
      PARAMETER (WCS_RESTFRQ  = 110)
      PARAMETER (WCS_RESTWAV  = 111)
      PARAMETER (WCS_NPV      = 112)
      PARAMETER (WCS_NPVMAX   = 113)
      PARAMETER (WCS_PV       = 114)
      PARAMETER (WCS_NPS      = 115)
      PARAMETER (WCS_NPSMAX   = 116)
      PARAMETER (WCS_PS       = 117)
      PARAMETER (WCS_CD       = 118)
      PARAMETER (WCS_CROTA    = 119)
      PARAMETER (WCS_ALTLIN   = 120)
      PARAMETER (WCS_VELREF   = 121)

      PARAMETER (WCS_ALT      = 122)
      PARAMETER (WCS_COLNUM   = 123)
      PARAMETER (WCS_COLAX    = 124)

      PARAMETER (WCS_CNAME    = 125)
      PARAMETER (WCS_CRDER    = 126)
      PARAMETER (WCS_CSYER    = 127)
      PARAMETER (WCS_DATEAVG  = 128)
      PARAMETER (WCS_DATEOBS  = 129)
      PARAMETER (WCS_EQUINOX  = 130)
      PARAMETER (WCS_MJDAVG   = 131)
      PARAMETER (WCS_MJDOBS   = 132)
      PARAMETER (WCS_OBSGEO   = 133)
      PARAMETER (WCS_RADESYS  = 134)
      PARAMETER (WCS_SPECSYS  = 135)
      PARAMETER (WCS_SSYSOBS  = 136)
      PARAMETER (WCS_VELOSYS  = 137)
      PARAMETER (WCS_ZSOURCE  = 138)
      PARAMETER (WCS_SSYSSRC  = 139)
      PARAMETER (WCS_VELANGL  = 140)
      PARAMETER (WCS_WCSNAME  = 141)

*     Codes for WCS data structure elements used by WCSGET (only).
      INTEGER   WCS_CEL, WCS_CUBEFACE, WCS_ERR, WCS_LAT, WCS_LATTYP,
     :          WCS_LIN, WCS_LNG, WCS_LNGTYP, WCS_NTAB, WCS_NWTB,
     :          WCS_SPC, WCS_SPEC, WCS_TAB, WCS_TYPES, WCS_WTB

      PARAMETER (WCS_NTAB     = 200)
      PARAMETER (WCS_NWTB     = 201)
      PARAMETER (WCS_TAB      = 202)
      PARAMETER (WCS_WTB      = 203)
      PARAMETER (WCS_LNGTYP   = 204)
      PARAMETER (WCS_LATTYP   = 205)
      PARAMETER (WCS_LNG      = 206)
      PARAMETER (WCS_LAT      = 207)
      PARAMETER (WCS_SPEC     = 208)
      PARAMETER (WCS_CUBEFACE = 209)
      PARAMETER (WCS_TYPES    = 210)
      PARAMETER (WCS_LIN      = 211)
      PARAMETER (WCS_CEL      = 212)
      PARAMETER (WCS_SPC      = 213)
      PARAMETER (WCS_ERR      = 214)

*     Flag bits for the AXES argument.
      INTEGER WCSSUB_CELESTIAL, WCSSUB_CUBEFACE, WCSSUB_LATITUDE,
     :        WCSSUB_LONGITUDE, WCSSUB_SPECTRAL, WCSSUB_STOKES

      PARAMETER (WCSSUB_LONGITUDE = 4096 +  1)
      PARAMETER (WCSSUB_LATITUDE  = 4096 +  2)
      PARAMETER (WCSSUB_CUBEFACE  = 4096 +  4)
      PARAMETER (WCSSUB_CELESTIAL = 4096 +  7)
      PARAMETER (WCSSUB_SPECTRAL  = 4096 +  8)
      PARAMETER (WCSSUB_STOKES    = 4096 + 16)

*     Error codes and messages.
      INTEGER   WCSERR_BAD_COORD_TRANS, WCSERR_BAD_CTYPE,
     :          WCSERR_BAD_PARAM, WCSERR_BAD_PIX, WCSERR_BAD_SUBIMAGE,
     :          WCSERR_BAD_WORLD, WCSERR_BAD_WORLD_COORD,
     :          WCSERR_ILL_COORD_TRANS, WCSERR_MEMORY,
     :          WCSERR_NON_SEPARABLE, WCSERR_NO_SOLUTION,
     :          WCSERR_NULL_POINTER, WCSERR_SINGULAR_MTX, WCSERR_SUCCESS

      PARAMETER (WCSERR_SUCCESS         =  0)
      PARAMETER (WCSERR_NULL_POINTER    =  1)
      PARAMETER (WCSERR_MEMORY          =  2)
      PARAMETER (WCSERR_SINGULAR_MTX    =  3)
      PARAMETER (WCSERR_BAD_CTYPE       =  4)
      PARAMETER (WCSERR_BAD_PARAM       =  5)
      PARAMETER (WCSERR_BAD_COORD_TRANS =  6)
      PARAMETER (WCSERR_ILL_COORD_TRANS =  7)
      PARAMETER (WCSERR_BAD_PIX         =  8)
      PARAMETER (WCSERR_BAD_WORLD       =  9)
      PARAMETER (WCSERR_BAD_WORLD_COORD = 10)
      PARAMETER (WCSERR_NO_SOLUTION     = 11)
      PARAMETER (WCSERR_BAD_SUBIMAGE    = 12)
      PARAMETER (WCSERR_NON_SEPARABLE   = 13)

      CHARACTER WCS_ERRMSG(0:13)*80
      COMMON /WCS_DATA/ WCS_ERRMSG
