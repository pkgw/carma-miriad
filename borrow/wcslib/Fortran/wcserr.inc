*=======================================================================
*
* WCSLIB 4.18 - an implementation of the FITS WCS standard.
* Copyright (C) 1995-2013, Mark Calabretta
*
* This file is part of WCSLIB.
*
* WCSLIB is free software: you can redistribute it and/or modify it
* under the terms of the GNU Lesser General Public License as published
* by the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* WCSLIB is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
* License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with WCSLIB.  If not, see http://www.gnu.org/licenses.
*
* Direct correspondence concerning WCSLIB to mark@calabretta.id.au
*
* Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
* http://www.atnf.csiro.au/people/Mark.Calabretta
* $Id$
*=======================================================================

*     Functions.
      EXTERNAL  WCSERR_ENABLE, WCSERR_GET, WCSERR_GTC, WCSERR_GTI,
     :          WCSERR_PRT
      INTEGER   WCSERR_ENABLE, WCSERR_GET, WCSERR_GTC, WCSERR_GTI,
     :          WCSERR_PRT

*     Length of the WCSERR data structure (INTEGER array) on 64-bit
*     machines.  Only needs to be 44 on 32-bit machines.
      INTEGER   ERRLEN
      PARAMETER (ERRLEN = 46)

*     Codes for WCSERR data structure elements used by WCSERR_GET.
      INTEGER   WCSERR_FILE, WCSERR_FUNCTION, WCSERR_LINE_NO,
     :          WCSERR_MSG, WCSERR_STATUS

      PARAMETER (WCSERR_STATUS   = 200)
      PARAMETER (WCSERR_LINE_NO  = 201)
      PARAMETER (WCSERR_FUNCTION = 202)
      PARAMETER (WCSERR_FILE     = 203)
      PARAMETER (WCSERR_MSG      = 204)

*     WCSERR message length.
      INTEGER   WCSERR_MSG_LENGTH
      PARAMETER (WCSERR_MSG_LENGTH = 160)
