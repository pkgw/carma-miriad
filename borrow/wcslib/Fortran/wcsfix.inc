*=======================================================================
*
* WCSLIB 4.18 - an implementation of the FITS WCS standard.
* Copyright (C) 1995-2013, Mark Calabretta
*
* This file is part of WCSLIB.
*
* WCSLIB is free software: you can redistribute it and/or modify it
* under the terms of the GNU Lesser General Public License as published
* by the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* WCSLIB is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
* License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with WCSLIB.  If not, see http://www.gnu.org/licenses.
*
* Direct correspondence concerning WCSLIB to mark@calabretta.id.au
*
* Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
* http://www.atnf.csiro.au/people/Mark.Calabretta
* $Id$
*=======================================================================

*     Functions.
      EXTERNAL  CDFIX, CELFIX, CYLFIX, DATFIX, SPCFIX, UNITFIX, WCSFIX
      INTEGER   CDFIX, CELFIX, CYLFIX, DATFIX, SPCFIX, UNITFIX, WCSFIX

*     Flag bits for the RELAX argument.
      INTEGER   WCSFIX_CD, WCSFIX_CEL, WCSFIX_CYL, WCSFIX_DAT,
     :          WCSFIX_NWCS, WCSFIX_SPC, WCSFIX_UNIT

      PARAMETER (WCSFIX_CD   = 1)
      PARAMETER (WCSFIX_DAT  = 2)
      PARAMETER (WCSFIX_UNIT = 3)
      PARAMETER (WCSFIX_CEL  = 4)
      PARAMETER (WCSFIX_SPC  = 5)
      PARAMETER (WCSFIX_CYL  = 6)
      PARAMETER (WCSFIX_NWCS = 6)

*     Error codes and messages.
      INTEGER   FIXERR_BAD_COORD_TRANS, FIXERR_BAD_CORNER_PIX,
     :          FIXERR_BAD_CTYPE, FIXERR_BAD_PARAM,
     :          FIXERR_ILL_COORD_TRANS, FIXERR_MEMORY, FIXERR_NO_CHANGE,
     :          FIXERR_NO_REF_PIX_COORD, FIXERR_NO_REF_PIX_VAL,
     :          FIXERR_NULL_POINTER, FIXERR_SINGULAR_MTX, FIXERR_SUCCESS

      PARAMETER (FIXERR_NO_CHANGE        = -1)
      PARAMETER (FIXERR_SUCCESS          =  0)
      PARAMETER (FIXERR_NULL_POINTER     =  1)
      PARAMETER (FIXERR_MEMORY           =  2)
      PARAMETER (FIXERR_SINGULAR_MTX     =  3)
      PARAMETER (FIXERR_BAD_CTYPE        =  4)
      PARAMETER (FIXERR_BAD_PARAM        =  5)
      PARAMETER (FIXERR_BAD_COORD_TRANS  =  6)
      PARAMETER (FIXERR_ILL_COORD_TRANS  =  7)
      PARAMETER (FIXERR_BAD_CORNER_PIX   =  8)
      PARAMETER (FIXERR_NO_REF_PIX_COORD =  9)
      PARAMETER (FIXERR_NO_REF_PIX_VAL   = 10)

      CHARACTER WCSFIX_ERRMSG(0:10)*80
      COMMON /WCSFIX_DATA/ WCSFIX_ERRMSG
