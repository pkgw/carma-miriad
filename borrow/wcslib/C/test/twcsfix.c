/*============================================================================

  WCSLIB 4.18 - an implementation of the FITS WCS standard.
  Copyright (C) 1995-2013, Mark Calabretta

  This file is part of WCSLIB.

  WCSLIB is free software: you can redistribute it and/or modify it under the
  terms of the GNU Lesser General Public License as published by the Free
  Software Foundation, either version 3 of the License, or (at your option)
  any later version.

  WCSLIB is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
  more details.

  You should have received a copy of the GNU Lesser General Public License
  along with WCSLIB.  If not, see http://www.gnu.org/licenses.

  Direct correspondence concerning WCSLIB to mark@calabretta.id.au

  Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
  http://www.atnf.csiro.au/people/Mark.Calabretta
  $Id$
*=============================================================================
*
* twcsfix tests the translation routines for non-standard WCS keyvalues, the
* wcsfix() suite, and the spectral coordinate translation routine wcssptr().
*
*---------------------------------------------------------------------------*/

#include <stdio.h>
#include <string.h>

#include <wcs.h>
#include <wcserr.h>
#include <wcsfix.h>
#include <wcsprintf.h>
#include <wcsunits.h>


void parser(struct wcsprm *);

const int NAXIS = 3;
const double CRPIX[3] =  {90.0, 90.0, 1.0};
const double PC[3][3] = {{ 1.0,  0.0, 0.0},
                         { 0.0,  1.0, 0.0},
                         { 0.0,  0.0, 1.0}};
const double CDELT[3] =  {-1.0, 1.0, 19.68717093222};

char CUNIT[3][9] = {"ARCSEC", "ARCSEC", "KM/SEC"};

/* N.B. non-standard. */
char CTYPE[3][9] = {"RA---NCP", "DEC--NCP", "FELO-HEL"};

const double CRVAL[3] = {265.6220947090*3600.0, -28.98849996030*3600.0,
                         5569.27104};
const double RESTFRQ = 1.42040575e9;
const double RESTWAV = 0.0;

/* N.B. non-standard, corresponding to MJD 35884.04861111 */
const char DATEOBS[] = "1957/02/15 01:10:00";

/* For testing spcfix(). */
const int  VELREF = 2;
const char SPECSYS[] = "BARYCENT";

int main()

{
  char ctypeS[9];
  int i, stat[NWCSFIX], status;
  struct wcsprm wcs;
  struct wcserr info[NWCSFIX];

  wcsprintf("Testing WCSLIB translator for non-standard usage (twcsfix.c)\n"
          "------------------------------------------------------------\n\n");

  wcs.flag = -1;
  parser(&wcs);

  /* Note: to print the unfixed wcsprm struct using wcsprt() the struct
     would first have to be initialized by wcsset().  However, if the struct
     contains non-standard keyvalues then wcsset() will either fix them
     itself or else fail (e.g. for non-standard units).  Thus, in general,
     wcsprt() cannot be used to print the unmodified struct. */

  /* Fix non-standard WCS keyvalues. */
  wcserr_enable(1);
  status = wcsfixi(7, 0, &wcs, stat, info);
  wcsprintf("wcsfix status returns: (");
  for (i = 0; i < NWCSFIX; i++) {
    wcsprintf(i ? ", %d" : "%d", stat[i]);
  }
  wcsprintf(")\n");

  for (i = 0; i < NWCSFIX; i++) {
    if (info[i].status < -1 || 0 < info[i].status) {
      wcsprintf("\n");
      wcserr_prt(info+i, 0x0);
    }
  }

  if (status) {
    wcsprintf("\nwcsfix error %d", status);
    return 1;
  }

  /* Extract information from the FITS header. */
  if (wcsset(&wcs)) {
    wcsprintf("\n");
    wcserr_prt(wcs.err, 0x0);
  }

  wcsprintf("\n");
  wcsprt(&wcs);
  wcsprintf("\n------------------------------------"
            "------------------------------------\n");

  /* Should now have a 'VOPT-F2W' axis, translate it to frequency. */
  strcpy(ctypeS, "FREQ-???");
  i = -1;
  if (wcssptr(&wcs, &i, ctypeS)) {
    wcserr_prt(wcs.err, 0x0);
    return 1;
  }

  if (wcsset(&wcs)) {
    wcserr_prt(wcs.err, 0x0);
    return 1;
  }

  wcsprt(&wcs);

  wcsfree(&wcs);

  return 0;
}

/*--------------------------------------------------------------------------*/

void parser(wcs)

struct wcsprm *wcs;

{
  int i, j;
  double *pcij;

  /* In practice a parser would read the FITS header until it encountered  */
  /* the NAXIS keyword which must occur near the start, before any of the  */
  /* WCS keywords.  It would then use wcsini() to allocate memory for      */
  /* arrays in the wcsprm struct and set default values.  In this          */
  /* simulation the header keyvalues are set as global variables.          */
  wcsnpv(2);
  wcsini(1, NAXIS, wcs);


  /* Now the parser scans the FITS header, identifying WCS keywords and    */
  /* loading their values into the appropriate elements of the wcsprm      */
  /* struct.                                                               */

  for (j = 0; j < NAXIS; j++) {
    wcs->crpix[j] = CRPIX[j];
  }

  pcij = wcs->pc;
  for (i = 0; i < NAXIS; i++) {
    for (j = 0; j < NAXIS; j++) {
      *(pcij++) = PC[i][j];
    }
  }

  for (i = 0; i < NAXIS; i++) {
    wcs->cdelt[i] = CDELT[i];
  }

  for (i = 0; i < NAXIS; i++) {
    strcpy(wcs->cunit[i], &CUNIT[i][0]);
  }

  for (i = 0; i < NAXIS; i++) {
    strcpy(wcs->ctype[i], &CTYPE[i][0]);
  }

  for (i = 0; i < NAXIS; i++) {
    wcs->crval[i] = CRVAL[i];
  }

  wcs->restfrq = RESTFRQ;
  wcs->restwav = RESTWAV;

  wcs->pv[0].i = -1;
  wcs->pv[0].m = -1;
  wcs->pv[0].value = -1.0;
  wcs->npv = 1;

  wcs->velref  = VELREF;
  strcpy(wcs->dateobs, DATEOBS);
  strcpy(wcs->specsys, SPECSYS);

  return;
}
