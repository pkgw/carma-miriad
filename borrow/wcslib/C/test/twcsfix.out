Testing WCSLIB translator for non-standard usage (twcsfix.c)
------------------------------------------------------------

wcsfix status returns: (-1, 0, 0, 0, 0, -1)

INFORMATIVE message from datfix() at line 402 of file wcsfix.c:
Changed '1957/02/15 01:10:00' to '1957-02-15T01:10:00'.

INFORMATIVE message from unitfix() at line 437 of file wcsfix.c:
Changed units: 'ARCSEC' -> 'arcsec', 'ARCSEC' -> 'arcsec', 'KM/SEC' -> 'km/s'.

INFORMATIVE message from spcfix() at line 484 of file wcsfix.c:
Changed CTYPE3 from 'FELO-HEL' to 'VOPT-F2W'.

       flag: 137
      naxis: 3
      crpix: 0x<address>
               90           90           1         
         pc: 0x<address>
    pc[0][]:   1            0            0         
    pc[1][]:   0            1            0         
    pc[2][]:   0            0            1         
      cdelt: 0x<address>
              -0.00027778   0.00027778   19687     
      crval: 0x<address>
               265.62      -28.988       5.5693e+06
      cunit: 0x<address>
             "deg"
             "deg"
             "m/s"
      ctype: 0x<address>
             "RA---SIN"
             "DEC--SIN"
             "VOPT-F2W"
    lonpole: 180.000000
    latpole: -28.988500
    restfrq: 1420405750.000000
    restwav: 0.000000
        npv: 3
     npvmax: 3
         pv: 0x<address>
              -1  -1  -1         
               2   1   0         
               2   2  -1.8049    
        nps: 0
     npsmax: 8
         ps: 0x<address>
         cd: 0x<address>
    cd[0][]:   0            0            0         
    cd[1][]:   0            0            0         
    cd[2][]:   0            0            0         
      crota: 0x<address>
               0            0            0         
     altlin: 0
     velref: 2
        alt: ' '
     colnum: 0
      colax: 0x<address>
                 0      0      0
    wcsname: UNDEFINED
      cname: 0x<address>
             UNDEFINED
             UNDEFINED
             UNDEFINED
      crder: 0x<address>
             UNDEFINED     UNDEFINED     UNDEFINED   
      csyer: 0x<address>
             UNDEFINED     UNDEFINED     UNDEFINED   
    radesys: UNDEFINED
    equinox: UNDEFINED
    specsys: "BARYCENT"
    ssysobs: UNDEFINED
    velosys: UNDEFINED
    ssyssrc: UNDEFINED
    zsource: UNDEFINED
     obsgeo: UNDEFINED     UNDEFINED     UNDEFINED     
    dateobs: "1957-02-15T01:10:00"
    dateavg: UNDEFINED
     mjdobs: 35884.048611
     mjdavg: UNDEFINED
       ntab: 0
        tab: 0x0
       nwtb: 0
        wtb: 0x0
      types: 0x<address>
            2200 2201 3300
     lngtyp: "RA"
     lattyp: "DEC"
        lng: 0
        lat: 1
       spec: 2
   cubeface: -1
        err: 0x0
        lin: (see below)
        cel: (see below)
        spc: (see below)
     m_flag: 137
    m_naxis: 3
    m_crpix: 0x<address>  (= crpix)
       m_pc: 0x<address>  (= pc)
    m_cdelt: 0x<address>  (= cdelt)
    m_crval: 0x<address>  (= crval)
    m_cunit: 0x<address>  (= cunit)
    m_ctype: 0x<address>  (= ctype)
       m_pv: 0x<address>  (= pv)
       m_ps: 0x<address>  (= ps)
       m_cd: 0x<address>  (= cd)
    m_crota: 0x<address>  (= crota)

    m_colax: 0x<address>  (= colax)
    m_cname: 0x<address>  (= cname)
    m_crder: 0x<address>  (= crder)
    m_csyer: 0x<address>  (= csyer)
      m_tab: 0x0  (= tab)
      m_wtb: 0x0  (= wtb)

   lin.*
       flag: 137
      naxis: 3
      crpix: 0x<address>
               90           90           1         
         pc: 0x<address>
    pc[0][]:   1            0            0         
    pc[1][]:   0            1            0         
    pc[2][]:   0            0            1         
      cdelt: 0x<address>
              -0.00027778   0.00027778   19687     
      unity: 1
        err: 0x0
     piximg: (nil)
     imgpix: (nil)
     m_flag: 0
    m_naxis: 0
    m_crpix: 0x0
       m_pc: 0x0
    m_cdelt: 0x0

   cel.*
      flag: 137
     offset: 0
       phi0:  0.000000
     theta0: 90.000000
       ref:   265.62      -28.988       180         -28.988    
       prj: (see below)
     euler:   265.62       118.99       180         -0.48463      0.87472   
    latpreq: 0 (not required)
     isolat: 0
        err: 0x0

   prj.*
       flag: 105
       code: "SIN"
         r0: 57.295780
         pv: (0)
              0           -1.8049    
       phi0:  0.000000
     theta0: 90.000000
     bounds: 1

       name: "orthographic/synthesis"
   category: 1 (zenithal)
    pvrange: 102
  simplezen: 0
  equiareal: 0
  conformal: 0
     global: 0
  divergent: 0
         x0: 0.000000
         y0: 0.000000
        err: 0x0
        w[]:   0.017453     3.2577       4.2577       2.2577       0         
               0            0            0            0            0         
          m: 0
          n: 0
     prjx2s: 0x<address>
     prjs2x: 0x<address>

   spc.*
       flag: 111
       type: "VOPT"
       code: "F2W"
      crval:  5.569e+06 
    restfrq: 1420405750.000000
    restwav: 0.000000
         pv: (not used)
          w:   0.2111       1.395e+09   -4.567       (remainder unused)
    isGrism: 0
        err: 0x0
     spxX2P: 0x<address>
     spxP2S: 0x<address>
     spxS2P: 0x<address>
     spxP2X: 0x<address>

------------------------------------------------------------------------
       flag: 137
      naxis: 3
      crpix: 0x<address>
               90           90           1         
         pc: 0x<address>
    pc[0][]:   1            0            0         
    pc[1][]:   0            1            0         
    pc[2][]:   0            0            1         
      cdelt: 0x<address>
              -0.00027778   0.00027778  -89906     
      crval: 0x<address>
               265.62      -28.988       1.3945e+09
      cunit: 0x<address>
             "deg"
             "deg"
             "Hz"
      ctype: 0x<address>
             "RA---SIN"
             "DEC--SIN"
             "FREQ"
    lonpole: 180.000000
    latpole: -28.988500
    restfrq: 1420405750.000000
    restwav: 0.000000
        npv: 3
     npvmax: 3
         pv: 0x<address>
              -1  -1  -1         
               2   1   0         
               2   2  -1.8049    
        nps: 0
     npsmax: 8
         ps: 0x<address>
         cd: 0x<address>
    cd[0][]:   0            0            0         
    cd[1][]:   0            0            0         
    cd[2][]:   0            0            0         
      crota: 0x<address>
               0            0            0         
     altlin: 0
     velref: 2
        alt: ' '
     colnum: 0
      colax: 0x<address>
                 0      0      0
    wcsname: UNDEFINED
      cname: 0x<address>
             UNDEFINED
             UNDEFINED
             UNDEFINED
      crder: 0x<address>
             UNDEFINED     UNDEFINED     UNDEFINED   
      csyer: 0x<address>
             UNDEFINED     UNDEFINED     UNDEFINED   
    radesys: UNDEFINED
    equinox: UNDEFINED
    specsys: "BARYCENT"
    ssysobs: UNDEFINED
    velosys: UNDEFINED
    ssyssrc: UNDEFINED
    zsource: UNDEFINED
     obsgeo: UNDEFINED     UNDEFINED     UNDEFINED     
    dateobs: "1957-02-15T01:10:00"
    dateavg: UNDEFINED
     mjdobs: 35884.048611
     mjdavg: UNDEFINED
       ntab: 0
        tab: 0x0
       nwtb: 0
        wtb: 0x0
      types: 0x<address>
            2200 2201 3000
     lngtyp: "RA"
     lattyp: "DEC"
        lng: 0
        lat: 1
       spec: 2
   cubeface: -1
        err: 0x0
        lin: (see below)
        cel: (see below)
        spc: (see below)
     m_flag: 137
    m_naxis: 3
    m_crpix: 0x<address>  (= crpix)
       m_pc: 0x<address>  (= pc)
    m_cdelt: 0x<address>  (= cdelt)
    m_crval: 0x<address>  (= crval)
    m_cunit: 0x<address>  (= cunit)
    m_ctype: 0x<address>  (= ctype)
       m_pv: 0x<address>  (= pv)
       m_ps: 0x<address>  (= ps)
       m_cd: 0x<address>  (= cd)
    m_crota: 0x<address>  (= crota)

    m_colax: 0x<address>  (= colax)
    m_cname: 0x<address>  (= cname)
    m_crder: 0x<address>  (= crder)
    m_csyer: 0x<address>  (= csyer)
      m_tab: 0x0  (= tab)
      m_wtb: 0x0  (= wtb)

   lin.*
       flag: 137
      naxis: 3
      crpix: 0x<address>
               90           90           1         
         pc: 0x<address>
    pc[0][]:   1            0            0         
    pc[1][]:   0            1            0         
    pc[2][]:   0            0            1         
      cdelt: 0x<address>
              -0.00027778   0.00027778  -89906     
      unity: 1
        err: 0x0
     piximg: (nil)
     imgpix: (nil)
     m_flag: 0
    m_naxis: 0
    m_crpix: 0x0
       m_pc: 0x0
    m_cdelt: 0x0

   cel.*
      flag: 137
     offset: 0
       phi0:  0.000000
     theta0: 90.000000
       ref:   265.62      -28.988       180         -28.988    
       prj: (see below)
     euler:   265.62       118.99       180         -0.48463      0.87472   
    latpreq: 0 (not required)
     isolat: 0
        err: 0x0

   prj.*
       flag: 105
       code: "SIN"
         r0: 57.295780
         pv: (0)
              0           -1.8049    
       phi0:  0.000000
     theta0: 90.000000
     bounds: 1

       name: "orthographic/synthesis"
   category: 1 (zenithal)
    pvrange: 102
  simplezen: 0
  equiareal: 0
  conformal: 0
     global: 0
  divergent: 0
         x0: 0.000000
         y0: 0.000000
        err: 0x0
        w[]:   0.017453     3.2577       4.2577       2.2577       0         
               0            0            0            0            0         
          m: 0
          n: 0
     prjx2s: 0x<address>
     prjs2x: 0x<address>

   spc.*
       flag: 0
       type: "    "
       code: "   "
      crval: UNDEFINED
    restfrq: 0.000000
    restwav: 0.000000
         pv: (not used)
          w:   0            0            0           (remainder unused)
    isGrism: 0
        err: 0x0
     spxX2P: 0x0
     spxP2S: 0x0
     spxS2P: 0x0
     spxP2X: 0x0
